davinci-resolve and davinci-resolve-studio: fix panel support and add missing applications

httpdev opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This commit fixes #366384, #364991, #364646, and possibly (untested) #168789.

It also adds several applications that are bundled with Resolve but were not supported yet:
- DaVinci Control Panels Setup
- DaVinci Remote Monitor
- Blackmagic RAW Player
- Blackmagic RAW Speed Test
- External Python scripting

This required some refactoring, so I am submitting this PR for discussion and expect that it would require some changes before accepting.

Changes:
 - Adds a number of missing dependencies required by the new applications.
 - The binary libraries required for the panels to work are now unpacked from `share/panels/dvpanel-framework-linux-x86_64.tgz` into `lib` so that they can be found from within the FHS environment.
 - The FHS environment formerly directly executed `bin/resolve` via `runScript` (discarding command line arguments). This did not allow to run the other supplied applications in the same FHS, therefore it now functions as a wrapper that can be called with the command+arguments to run. For debugging, the package also has a `bin/davinci-resolve-*-fhs` command, which starts a shell in the FHS environment.
 - The main package is now created using `symlinkJoin` to have better control of which files are linked where. The main package only includes the application links in `bin/`, udev rules in `etc/udev/rules.d/`, and desktop items in `share/applications/`.
 - The `udev` rules now follow the same naming and priority as when installing Resolve on Non-NixOS. These files were formerly only mapped from the unpacked archive (`99-*.rules`), but in the Resolve install script they are actually modified and a new file is added for the dongle, which results in `75-davincipanel.rules`, `75-davincikb.rules`, and `75-sdx.rules` (new).
 - Desktop items for the new applications are added and the main desktop item now has `mimeTypes` set.
 - MIME types XML and icons are installed into the XDG directory structure
 - A `longDescription` is added to hint to the fact that the package needs to be added to `services.udev.packages` in order for the panels to work, which was not documented, and includes comments about scripting.

I have tested using a DaVinci Resolve Micro Panel and DaVinci Resolve Speed Editor, which work including updating the firmware using the Control Panels Setup application. I do not have access to panels with a display, so I am not sure if they require the supplied daemon, which would require additional setup.

External scripting works with DaVinci Resolve Studio. Icons are correctly displayed on GNOME, Resolve project files (.drp) are correctly opened by Resolve, and RAW file (.braw) open the RAW player (see the three clips that can be downloaded on [this page](https://www.blackmagicdesign.com/products/blackmagicraw) for testing).

I have tested the automatic update script and it still works as far as I can tell.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

httpdev commented on pkgs/by-name/da/davinci-resolve/package.nix:
```diff
@@ -149,8 +152,11 @@ let
+          mkdir -p $out/lib
+          tar xvf $out/share/panels/dvpanel-framework-linux-x86_64.tgz -C $out/lib
 
           mkdir -p $out/{configs,DolbyVision,easyDCP,Fairlight,GPUCache,logs,Media,"Resolve Disk Database",.crashreport,.license,.LUT}
```
fixed

httpdev replied:
fixed.

amarshall reviewed (changes requested):
Thanks for your contribution. Please read the [contributing guide](https://github.com/NixOS/nixpkgs/blob/d86ae899d2909c0899e4d3b29d90d5309771e77c/CONTRIBUTING.md), in particular the section on commit conventions. The commits here need to be renamed to match convention, as well as be atomic and semantic. The PR is a bit difficult to review as-is, as there are a lot of redundant commits, and not all commits are explained by their commit message, so it’s challenging to understand why some things are being changed.

I added some inline comments but stopped quickly as it was, as above, challenging to follow along.

TL;DR: Please reorganize so commits are smaller, discrete changes that are independently useful, rather than having commits make many unrelated changes. Also avoid having commits that only fix issues with earlier ones—instead amend-in the changes. There’s a lot of detail in the PR description—instead much of that should be in the discrete commit messages so it’s easy to follow and link the why to the what.

httpdev commented on pkgs/by-name/da/davinci-resolve/package.nix:
```diff
@@ -32,9 +34,10 @@
   davinci = (
     stdenv.mkDerivation rec {
-      pname = "davinci-resolve${lib.optionalString studioVariant "-studio"}";
+      pname = "davinci-resolve${lib.optionalString studioVariant "-studio"}-base";
```
This is mostly to be able to identify the derivations for the raw extracted package (-base) and the main (user-facing) package, which would otherwise both be called "<hash>-davinci-resolve-studio" with completely different directory structure. I'm not sure what the convention here is, but buildFHSEnv does a similar thing.

httpdev commented on pkgs/by-name/da/davinci-resolve/package.nix:
```diff
@@ -14,6 +14,8 @@
   bash,
   writeText,
+  writeShellScriptBin,
+  symlinkJoin,
```
changed to
symlinkJoin,
writeShellScriptBin,
writeText,

httpdev commented:
> Thanks for your contribution. Please read the [contributing guide](https://github.com/NixOS/nixpkgs/blob/d86ae899d2909c0899e4d3b29d90d5309771e77c/CONTRIBUTING.md), in particular the section on commit conventions. The commits here need to be renamed to match convention, as well as be atomic and semantic. The PR is a bit difficult to review as-is, as there are a lot of redundant commits, and not all commits are explained by their commit message, so it’s challenging to understand why some things are being changed.
> 
> I added some inline comments but stopped quickly as it was, as above, challenging to follow along.
> 
> TL;DR: Please reorganize so commits are smaller, discrete changes that are independently useful, rather than having commits make many unrelated changes. Also avoid having commits that only fix issues with earlier ones—instead amend-in the changes. There’s a lot of detail in the PR description—instead much of that should be in the discrete commit messages so it’s easy to follow and link the why to the what.

Thanks for the review. I understand, this is my first contribution to nixpkgs, so I'm not used to the conventions.

I'm not sure how to split into independent commits. There are basically two parts: supporting the other utilities, and fixing the panel udev issues. Supporting the other utilities requires separating out the FHS environment into a separate derivation, which needs some reorganization. Supporting the panel setup tool without fixing the panel udev issues doesn't make sense, and fixing the udev issues and then reorganizing them anyway in the next commit also doesn't seem useful.

httpdev committed:
davinci-resolve: refactor FHS environment

Refactors the FHS environment into a reusable wrapper to prepare for
supporting the currently unavailable tools supplied in the binary
distribution file. Adds a davinci-resolve-*-shell command which wraps a
shell inside the FHS for scripting and debugging. Adds corresponding
documentation in longDescription.

httpdev committed:
davinci-resolve: add udev rules for panels and hardware dongle

fixes #366384
fixes #364991
possible fix for #364646 and #168789 (untested)

The udev rules now follow the same naming and priority as when
installing Resolve on Non-NixOS. These files were formerly mapped from
the unpacked archive (99-*.rules), but in the Resolve install script
they are actually modified and a new file is added for the dongle, which
results in 75-davincipanel.rules, 75-davincikb.rules, and 75-sdx.rules
(new). Rules are in etc/udev/rules.d.

Expanded longDescription to hint users to the fact that the package
needs to be added to services.udev.packages in order for the panels to
work, which was not documented.

httpdev committed:
davinci-resolve: moved desktop item and icon to final package

httpdev committed:
davinci-resolve: fix desktop item

This adds the mimeTypes, startupNotify and terminal settings to the
desktop item as in the original binary distribution.

httpdev committed:
davinci-resolve: add control panels setup application

This adds the control panels setup dependencies, wrapper, and desktop
item.

httpdev committed:
davinci-resolve: add remote monitor

This adds the remote monitor wrapper and desktop item.

httpdev committed:
davinci-resolve: add BlackMagic RAW Speed Test

This adds the speed test application wrapper and desktop item.

httpdev commented:
Hello, I have restructured the pull request. When requesting further changes, I kindly ask to consider that debugging the hardware issues and missing dependencies alone was more than 30 hours of work.

httpdev committed:
davinci-resolve: remove redundant use of optionalString

This commit improves code readability by replacing expressions involving
optionalString with ${davinci.pname} and ${product}.

httpdev committed:
davinci-resolve: support different LD_LIBRARY_PATHs for applications

The applications shipped in the binary distribution come with their own
sets of dynamic libraries. This commit introduces two additional options
in mkWrapper that allow to specify the library path and QT plugin
directory for each application individually instead of using the
libraries provided by the main executable.

httpdev committed:
davinci-resolve: add RAW player

This commit adds a wrapper and desktop items to support the Blackmagic
Design RAW Player.

httpdev committed:
davinci-resolve: add mime types

This commit adds the .xml files with the MIME type definitions and
corresponding icons for DaVinci Resolve and Blackmagic RAW files.

httpdev committed:
davinci-resolve: add python scripting support

This commit adds Python scripting support. It sets the required paths in
the FHS environment and creates a "davinci-resolve-*-python" wrapper. It
also adds an additional "pythonPackage" input which can be used to
override the Python package and documentation. Note that external
scripting appears to be supported only by the non-free Studio version.

httpdev commented:
Nevermind, I have found some more time to work on this. It was a bit tricky because the binary distribution is a mess - it consists of several binaries which each come with their own (incompatible) .so libraries. This required a slight modification to the wrapper so they can each live in their own happy space.

The RAW player is now working and all MIME types are correctly added with their icons. I have also added a command `davinci-resolve-studio-python` which runs Python inside the FHS environment and allows external scripting. An example is in the longDescription.

flexiondotorg reviewed (approved):
I've tested this locally, and it is a huge improvement.

httpdev commented:
@amarshall @jshcmpbll Hello, any chance of having another look at this/getting it merged? I have taken some time to structure the PR clearly and it already has another independent approval.

httpdev committed:
davinci-resolve: clean up formatting

httpdev commented:
rebased onto current master, checked that it builds with current Resolve 20.1 & cleaned up formatting using nixfmt.
